include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+                          "variable (and hence we cannot take its derivative "
+                          "directly)."));
+        Assert(row_n != dependent_direction,
+               ExcMessage("We cannot differentiate with respect to the variable "
+                          "that is assumed to be dependent."));
+
+        const Point<spacedim> manifold_point = f(center);
+        const Tensor<1, spacedim> stencil_value = cross_stencil<structdim>
+                                                  ({row_n, dependent_direction},
+                                                   center,
+                                                   step,
+                                                   f);
+        double entry = 0.0;
+        for (unsigned int dim_n = 0; dim_n < spacedim; ++dim_n)
+          entry += -2.0*(p0[dim_n] - manifold_point[dim_n])*stencil_value[dim_n];
+        return entry;
+      }
+
+      /**
+       * Project onto a flat object. This is more accurate than the general
+       * algorithm in project_to_object.
+       */
+      template <typename Iterator, int spacedim, int structdim>
+      Point<spacedim>
+      project_to_flat_object (const Iterator        &object,
+                              const Point<spacedim> &trial_point)
+      {
+        // let's look at this for simplicity for a quad (structdim==2) in a space with
+        // spacedim>2 (notate trial_point by y): all points on the surface are
+        // given by
+        //   x(\xi) = sum_i v_i phi_x(\xi)
+        // where v_i are the vertices of the quad, and \xi=(\xi_1,\xi_2) are the
+        // reference coordinates of the quad. so what we are trying to do is find
+        // a point x on the surface that is closest to the point y. there are
+        // different ways to solve this problem, but in the end it's a nonlinear
+        // problem and we have to find reference coordinates \xi so that J(\xi) =
+        // 1/2 || x(\xi)-y ||^2 is minimal. x(\xi) is a function that is
+        // structdim-linear in \xi, so J(\xi) is a polynomial of degree 2*structdim that we'd
+        // like to minimize. unless structdim==1, we'll have to use a Newton method to
+        // find the answer. This leads to the following formulation of Newton
+        // steps:
+        //
+        // Given \xi_k, find \delta\xi_k so that
+        //   H_k \delta\xi_k = - F_k
+        // where H_k is an approximation to the second derivatives of J at \xi_k,
+        // and F_k is the first derivative of J.  We'll iterate this a number of
+        // times until the right hand side is small enough. As a stopping
+        // criterion, we terminate if ||\delta\xi||<eps.
+        //
+        // As for the Hessian, the best choice would be
+        //   H_k = J''(\xi_k)
+        // but we'll opt for the simpler Gauss-Newton form
+        //   H_k = A^T A
+        // i.e.
+        //   (H_k)_{nm} = \sum_{i,j} v_i*v_j *
```

Comment:
this is `v_i \otimes v_j`, right?

Reply:
That is a dot product of the vertices; I don't think we use `otimes` for dot products. Do we?

Reply:
Correct -- otimes is for the outer product, not the dot product. I now also see that this of course matches the code below.